### DISCUSSION TITLE
Custom dockerfile from prebuilt nvidia-pytroch

### DISCUSSION THREAD

**Discussion by skol101** [espnet/espnet#3909] [Q&A]
I want to minimize the image size (espnet built image is about 14Gb), thus I'm thinking about using pre-build Nvidia pytorch cuda runtime image https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch .

If pytorch is installed already, I assume that ESPnet installation won't re-install PyTorch again?

**Comment by sw005320** [comment]
@Fhrozen, what do you think?
I think it sounds reasonable.

**Reply by skol101** [reply]
Is it possible to install ESPNet without Kaldi and Conda (since it's already installed in the Nvidia Pytorch container)? 

My first run wasn't successfull
```
 => [1/7] FROM nvcr.io/nvidia/pytorch:21.12-py3                           0.0s
 => CACHED [2/7] RUN git clone https://github.com/espnet/espnet           0.0s
 => CACHED [3/7] WORKDIR /espnet/tools                                    0.0s
 => [4/7] RUN ./setup_anaconda.sh /opt/conda espnet 3.8                  30.7s
 => ERROR [5/7] RUN make                                                  5.7s 
------                                                                         
 > [5/7] RUN make:                                                             
#8 0.389 test -f kaldi/egs/wsj/s5/utils/parse_options.sh || git clone --depth 1 https://github.com/kaldi-asr/kaldi                                            
#8 0.400 Cloning into 'kaldi'...         
Pytorch=1.10.1 binary for CUDA=11.5 is not provided. 11.3 11.1 10.2 are supported.
#8 5.590 make: *** [Makefile:98: pytorch.done] Error 1

```

**Reply by Fhrozen** [reply]
I will try to check. There are so many libraries that support not only espnet 2, but also espnet that makes the container have that size. I think I can separate containers for both versions of espnet

**Reply by Fhrozen** [reply]
Use RUN make CUDA_VERSION=11.1

**Reply by skol101** [reply]
Cheers, looks like I have to use 21.06-py3 which has required CUDA and Pytorch versions. Also, 21.06 is 15Gb big (uncompressed), so I guess I'd better just pack what's needed myself, though not sure how much can be saved.

**Reply by skol101** [reply]
Managed to build the image by
```
FROM nvcr.io/nvidia/pytorch:21.11-py3
WORKDIR /
# Install espnet
RUN git clone https://github.com/espnet/espnet
WORKDIR /espnet/tools
ENV CONDA_TOOLS_DIR=/opt/conda
RUN ./setup_anaconda.sh ${CONDA_TOOLS_DIR} espnet 3.8
RUN touch pytorch.done
RUN make showenv CUDA_VERSION=11.1
WORKDIR /

```

After exec into container
```
root@c58a38c9d95b:/espnet/tools# python check_install.py
[x] python=3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:59:51)  [GCC 9.4.0]

Python modules:
[x] torch=1.11.0a0+b6df043
[x] torch cuda=11.5
[x] torch cudnn=8300
[x] torch nccl
[ ] chainer
[x] cupy=9.3.0
[x] cupy nccl

**Reply by skol101** [reply]
Here's another way to make custom local docker containers https://github.com/veritas9872/PyTorch-Universal-Docker-Template . Supposedly, they should compute faster than standard NGC images.
